cg/services/order_validation_service/utils.py

```diff
@@ -19,6 +20,14 @@ def apply_order_validation(rules: list[Callable], order: Order, store: Store) ->
     return errors
 
 
+def apply_case_validation(rules: list[Callable], order: Order, store: Store) -> list[CaseError]:
```

Comment:
Is there a point to having "apply_case_validation" as a new method and not reuse the others?

Reply:
The existing division is just based on the different error types the different rules produce. I guess this is the first rule that generates a pure `CaseError` and as such it needs a separate function.

Reply:
There might be a better way to do it but it is not apparent to me atm, maybe we can revisit it later?

Reply:
I was thinking that we could have a "apply_validation" and type hint it with a very general error, and then the list of rules will in the end be the only thing dictating the return type. Given that Python doesn't enforce these type hints (as far as I know), that is the current behaviour either way.

Reply:
I don't follow 😞 But since we have this division of errors
```python
class ValidationErrors(BaseModel):
    order_errors: list[OrderError] | None = None
    case_errors: list[CaseError] | None = None
    sample_errors: list[SampleError] | None = None
    case_sample_errors: list[CaseSampleError] | None = None
```

I think it is necessary to have a separate `apply` for all of the different types of errors

Reply:
I just mean that the `apply` methods can take any of the rules and return any kind of error. The only difference between them are their names and the type hinted return type. To my knowledge we would get exactly the same behaviour from

![image](https://github.com/user-attachments/assets/62a00cd1-9fd2-4c60-b4e6-abeeabe8d8f7)

as from 

![image](https://github.com/user-attachments/assets/20aa7e1d-ef10-4658-aa6c-6e031245fee8)

although we of course get warnings when working in the code.

Reply:
Maybe we could rewrite the different errors so that we can treat them interchangeably but I don't see how. I'd rather we left it as is now and just have four roughly identical functions 😅

Reply:
I will leave it as is for now! We could probably have some parent error for it if we are desperate but I am fine as is also.